vcpkg{,-tool}: miscellaneous fixups

h7x4 opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

**Summary:**

- Moved most of the wrapper from vcpkg to vcpkg-tool directly (so you can rely on vcpkg-tool with an external VCPKG_ROOT and not have to repeat the patches)
- Ensure all of the arguments in the (now) vcpkg-tool wrapper is overridable, using envvars prefixed with `NIX` to communicate that these are nix specific.
- Patch some of the vcpkg cmake scripts, to conform the compiler detection to the executable names used in `pkgsCross`.
- Add some missing runtime deps for vcpkg-tool (these are listed as direct dependencies at https://github.com/microsoft/vcpkg/blob/101cc9a69a1061969caf4b73579a34873fdd60fe/scripts/bootstrap.sh#L88-L91 as well as https://github.com/microsoft/vcpkg/blob/101cc9a69a1061969caf4b73579a34873fdd60fe/scripts/azure-pipelines/linux/provision-image.sh#L27)

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

h7x4 committed:
vcpkg{,-tool}: move large part of the wrapper to vcpkg-tool

h7x4 committed:
vcpkg-tool: ensure opinionated config is overridable

- use overridable environment variables for input
- prefix those variables with "NIX_", to communicate that this is
  something specific to nixpkgs
- detect equal cli flags, and override envvars based on them
- add preferred path "$XDG_CACHE_HOME/vcpkg" before "$HOME/.vcpkg/root"
- set "--x-install-root"

Co-authored-by: Edgar B <39066502+Guekka@users.noreply.github.com>

h7x4 committed:
vcpkg: add quotations

h7x4 committed:
vcpkg: patch scripts to find arm cross-compilers automatically

h7x4 committed:
vcpkg-tool: add missing runtime deps

Guekka commented:
This looks good, thank you! I've been wanting to add XDG compliance for a while
I assume you have tested the changes?

Nitpick: I think `substituteInPlace` is recommended over sed

h7x4 commented:
> This looks good, thank you! I've been wanting to add XDG compliance for a while
> I assume you have tested the changes?

Admittedly, I have not tested with this exact patch. I created the patch out of a bunch of local overrides in a project. I'll try building it with overlayed changes from this PR.

> Nitpick: I think `substituteInPlace` is recommended over sed

Just to be clear, `substituteInPlace` does not allow for regexes. Would you be okay with unrolling `(as|gcc|g++)` to 3 similar copies of `--replace-fail`?

Guekka commented:
> Just to be clear, substituteInPlace does not allow for regexes. Would you be okay with unrolling (as|gcc|g++) to 3 similar copies of --replace-fail?

Oh, right. I only suggested it because I think it's common in nixpkgs. But actually, I think it might be more obvious this way, so let's keep it like this if you're fine with it

I'll approve once I get a chance to test it, should be pretty soon

h7x4 commented:
On a second note, this is a pretty breaking change since we now do not allow use of `${vcpkg-tool}/bin/vcpkg` with the 4 flags we are setting (I know I at least have used this as an escape hatch in the past, to avoid the overrides in `${vcpkg}/bin/vcpkg`). I think we should write a release note, and maybe document the use of these 4 variables somewhere.

I'm also open to do some kind of magic, where we detect the incoming flags from outside the wrappeer and drop the overrides accordingly. I think it would make for a more native-feeling, seamless experience. But I'm not very confident in my bash skills :sweat_smile: 

I'll mark the PR as draft in the meantime.

gracicot commented:
@h7x4 This is a breaking change indeed, but I'm open to not break too much if there's benefit. The magic flags seems okay to me. It's not too magic, it just let the users override default locations we set up for them.

Also if `vcpkg-tool` is wrapped by default, I would also provide `vcpkg-tool-unwrapped`, as with other build tools that have a wrapper

For example, I might type such command to manage a private registry:
```
vcpkg --x-builtin-ports-root=./ports --x-builtin-registry-versions-dir=./versions x-add-version --all --verbose
```

Having a unwrapped version would make sense for such scripts

h7x4 committed:
vcpkg: add 'doWrap' flag

h7x4 commented:
I have added a new part to the wrapper that will loop over the args and detect user-provided versions of the args we override. This should make the UX better, while still ensuring the tool doesn't write to the nix store.

Guekka commented on pkgs/by-name/vc/vcpkg-tool/package.nix:
```diff
@@ -47,23 +49,100 @@ stdenv.mkDerivation (finalAttrs: {
-    wrapProgram $out/bin/vcpkg --prefix PATH ${lib.makeBinPath runtimeDeps}
+    #!${runtimeShell}
+
+    NIX_VCPKG_WRITABLE_PATH=''${NIX_VCPKG_WRITABLE_PATH:-''${XDG_CACHE_HOME+"$XDG_CACHE_HOME/vcpkg"}}
+    NIX_VCPKG_WRITABLE_PATH=''${NIX_VCPKG_WRITABLE_PATH:-''${HOME+"$HOME/.vcpkg/root"}}
+    NIX_VCPKG_WRITABLE_PATH=''${NIX_VCPKG_WRITABLE_PATH:-''${TMP}}
+    NIX_VCPKG_WRITABLE_PATH=''${NIX_VCPKG_WRITABLE_PATH:-'/tmp'}
```
For someone who doesn't know well bash, this was not easy to grasp. What do you think of something like this? Longer, but more explicit
```suggestion
    get_vcpkg_path() {
      if [[ -n $NIX_VCPKG_WRITABLE_PATH ]]; then
          echo "$NIX_VCPKG_WRITABLE_PATH"
      elif [[ -n $XDG_CACHE_HOME ]]; then
          echo "$XDG_CACHE_HOME/vcpkg"
      elif [[ -n $HOME ]]; then
          echo "$HOME/.vcpkg/root"
      elif [[ -n $TMP ]]; then
          echo "$TMP"
      else
          echo "/tmp"
      fi
    }

    NIX_VCPKG_WRITABLE_PATH=$(get_vcpkg_path)
```

Guekka commented on pkgs/by-name/vc/vcpkg-tool/package.nix:
```diff
@@ -47,23 +49,100 @@ stdenv.mkDerivation (finalAttrs: {
+    ARGS=( "$@" )
+    FINAL_NONMODIFIED_ARGS=()
+
+    for (( i=0; i<''${#ARGS[@]}; i++ ));
+    do
+      case "''${ARGS[i]%%=*}" in
+        ${let
+          f = { arg, env, ... }: ''
+            '${arg}')
+              ${env}="''${ARGS[i]#*=}"
+              if [ "''$${env}" = '${arg}' ]; then
+                ${env}="''${ARGS[i+1]}"
+                ((i++))
+              fi
+              ;;
+          '';
+        in lib.concatMapStringsSep "\n" f argsWithDefault}
+        *)
+          FINAL_NONMODIFIED_ARGS+=(''${ARGS[i]})
+          ;;
+      esac
+    done
```
This logic is (necessarily) complex. I'm wondering if it's worth writing a test

h7x4 replied:
I think it would be nice both for testing and for debugging to provide an environment variable that will debug print the final cli flags in the wrapper before calling `vcpkg`. We could create a test where we set this envvar, and run the tool with a command such as `contact` or `--version`, and then verify that the resulting flags from the debug output looks sane.

What do you think?

Guekka replied:
Sounds good to me

h7x4 replied:
I've added a basic test. It should be able to report any syntactic failures of the wrapper in case the script suddenly doesn't run at all, as well as basic functionality errors.

Guekka reviewed (commented):
I think your script has the right idea

h7x4 committed:
vcpkg-tool: add debug-print functionality to wrapper

h7x4 committed:
vcpkg-tool: add basic test for wrapper

h7x4 committed:
vcpkg-tool: add meta.changelog

h7x4 committed:
vcpkg-tool: add h7x4 as maintainer

h7x4 committed:
vcpkg: add h7x4 as maintainer

Guekka commented on pkgs/by-name/vc/vcpkg-tool/package.nix:
```diff
@@ -47,31 +51,139 @@ stdenv.mkDerivation (finalAttrs: {
+
+    NIX_VCPKG_WRITABLE_PATH=$(get_vcpkg_path)
+
+    ${lib.concatMapStringsSep "\n" ({ env, default, ... }: ''${env}=''${${env}-"${default}"}'') argsWithDefault}
```
If I'm not mistaken, this expression is repeated thrice and should be extracted to a function

h7x4 replied:
We actually use it 4 places at this point. But I am not too sure whether it makes for a good abstraction.

The only thing common between the invocations is that we apply some function for all the arguments with defaults, the function returns a string, and then concatenate with `\n` between. The intent behind the functions are very different. While something like `applyFuncForArgsWithDefaultAndConcat` might save some space, it doesn't save any evaluation time, nor do I think it makes it much clearer to the reader.

But I'd be happy if we could find a better expression

h7x4 committed:
vcpkg-tool-unwrapped: init

h7x4 committed:
doc/packages/vcpkg: init

h7x4 commented:
@gracicot: added a `vcpkg-tool-unwrapped` attribute in top-level.

I've added a piece of documentation that will end up in the nixpkgs manual. Built with `nix build .#htmlDocs.nixpkgsManual.x86_64-linux` and verified that the formatting looks sane (but I'd like some verification that the content is reasonable as well)

I can't think of any more changes that I'd like to do before merging this, so I'll undraft the PR. Happy if I could get some more reviews :)

Guekka reviewed (approved):
Looks good to me, thank you!

gracicot commented:
Overlay triplets doesn't seem to work for me. I'm investigating

gracicot commented on pkgs/by-name/vc/vcpkg-tool/package.nix:
```diff
@@ -47,31 +51,139 @@ stdenv.mkDerivation (finalAttrs: {
+
+    exec -a "$0" "${placeholder "out"}/bin/.vcpkg-wrapped" \
+    ${lib.concatMapStringsSep "\n" ({ arg, env, ... }: "  " + ''${arg}="''$${env}" \'') argsWithDefault}
+      "$FINAL_NONMODIFIED_ARGS"
```
Arguments sent from `scripts/buildsystems/vcpkg.cmake` are not forwarded, leading to overlay triplets and target triplets to not work

h7x4 replied:
Thanks for looking into it. I'll have a closer look soon to see if we can make the proposed changes work somehow

gracicot replied:
Yes! I'm really looking forward those. It's a good cleanup that was overdue.

h7x4 commented on pkgs/by-name/vc/vcpkg/package.nix:
```diff
@@ -16,27 +16,29 @@ stdenvNoCC.mkDerivation (finalAttrs: {
+      --replace-fail "aarch64-linux-gnu-g++" "aarch64-unknown-linux-gnu-g++" \
+      --replace-fail "arm-linux-gnueabihf-as"  "armv7l-unknown-linux-gnueabihf-as" \
+      --replace-fail "arm-linux-gnueabihf-gcc" "armv7l-unknown-linux-gnueabihf-gcc" \
+      --replace-fail "arm-linux-gnueabihf-g++" "armv7l-unknown-linux-gnueabihf-g++"
```
While the binary names might not be nixpkgs-only, I believe they differ from most distros. Ubuntu at least uses the ones specified upstream. I think we should leave them be.

gracicot commented on pkgs/by-name/vc/vcpkg-tool/package.nix:
```diff
@@ -47,31 +51,139 @@ stdenv.mkDerivation (finalAttrs: {
+
+    export PATH="${lib.makeBinPath runtimeDeps}''${PATH:+":$PATH"}"
+
+    ARGS=( "$@" )
```
The problem seems to originate from here. I changed the script to output the values. Here's what I did:
```
ARGS=( "$@" )
echo "REALLY_ALL_ARGS $@"
echo "ARGS ${ARGS}"
```
Here's the output I see in my project:
```
REALLY_ALL_ARGS install --triplet wasm32-unknown-emscripten --vcpkg-root /Users/grc/Workspace/github/nixpkgs/result/share/vcpkg --x-wait-for-lock --x-manifest-root=/Users/grc/Workspace/software/pixmob/apps/visualizer/visualizer-web/packages/vtx --x-install-root=/Users/grc/Workspace/software/pixmob/apps/visualizer/visualizer-web/packages/vtx/build/dev-emscripten/vcpkg_installed --host-triplet=aarch64-darwin-clang --overlay-triplets=/Users/grc/Workspace/software/pixmob/apps/visualizer/visualizer-web/packages/vtx/triplets
ARGS install
```
Here `ARGS` should not be just `install`, it should be an array with all the args above, if I understood the scripts correctly.

h7x4 replied:
This seems to be an issue indeed. `$ARRAY` only yields the first element of the array, while `${ARRAY[@]}` yields all elements. I seem to have overlooked it for `FINAL_NONMODIFIED_ARGS`, making it so only the first of the non-special arguments are forwarded from the wrapper. I'll push a fix. Does it resolve the overlay issues as well?

gracicot replied:
The updated version seems to work as expected. I approve the changes!
